// botbuilder-teams/lib/ConversationUpdate/membersRemovedEvent.js
'use strict';
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var models = require('../models');
var TeamEventBase = require('./teamEventBase').TeamEventBase;
var TeamEventType = require('./teamEventBase').TeamEventType;
var MembersRemovedEvent = (function (_super) {
    __extends(MembersRemovedEvent, _super);
    function MembersRemovedEvent(membersRemoved, team, tenant) {
        var _this = _super.call(this, MembersRemovedEvent.eventType, team, tenant) || this;
        _this.membersRemoved = membersRemoved;
        return _this;
    }
    MembersRemovedEvent.prototype.mapper = function () {
        return {
            required: false,
            serializedName: 'MembersRemovedEvent',
            type: {
                name: 'Composite',
                className: 'MembersRemovedEvent',
                modelProperties: {
                    membersRemoved: {
                        required: false,
                        serializedName: 'membersRemoved',
                        type: {
                            name: 'Sequence',
                            element: {
                                required: false,
                                serializedName: 'ChannelAccountType',
                                type: {
                                    name: 'Composite',
                                    className: 'ChannelAccount'
                                }
                            }
                        }
                    }
                }
            }
        };
    };
    return MembersRemovedEvent;
}(TeamEventBase));
MembersRemovedEvent.eventType = TeamEventType.MembersRemoved;
module.exports = MembersRemovedEvent;
