// botbuilder-teams/lib/models/FileInfoCard.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var FileInfoCard = (function () {
    function FileInfoCard(session) {
        this.session = session;
        this.data = {
            contentType: "application/vnd.microsoft.teams.card.file.info",
            content: {}
        };
    }
    FileInfoCard.prototype.name = function (name) {
        this.data.name = name;
        return this;
    };
    FileInfoCard.prototype.contentUrl = function (url) {
        this.data.contentUrl = url;
        return this;
    };
    FileInfoCard.prototype.uniqueId = function (uniqueId) {
        this.data.content.uniqueId = uniqueId;
        return this;
    };
    FileInfoCard.prototype.fileType = function (fileType) {
        this.data.content.fileType = fileType;
        return this;
    };
    FileInfoCard.prototype.toAttachment = function () {
        return this.data;
    };
    FileInfoCard.fromFileUploadInfo = function (uploadInfo) {
        return new FileInfoCard()
            .name(uploadInfo.name)
            .contentUrl(uploadInfo.contentUrl)
            .uniqueId(uploadInfo.uniqueId)
            .fileType(uploadInfo.fileType);
    };
    return FileInfoCard;
}());
exports.FileInfoCard = FileInfoCard;
