// botbuilder-teams/lib/models/FileConsentCard.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var o365ConnectorCard_1 = require("./o365ConnectorCard");
var FileConsentCard = (function () {
    function FileConsentCard(session) {
        this.session = session;
        this.data = {
            contentType: "application/vnd.microsoft.teams.card.file.consent",
            content: {}
        };
    }
    FileConsentCard.prototype.name = function (name) {
        this.data.name = name;
        return this;
    };
    FileConsentCard.prototype.description = function (description) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        if (description) {
            this.data.content.description = o365ConnectorCard_1.fmtText(this.session, description, args);
        }
        else {
            delete this.data.content.description;
        }
        return this;
    };
    FileConsentCard.prototype.sizeInBytes = function (sizeInBytes) {
        if (sizeInBytes < 0) {
            throw new Error("sizeInBytes must be greater than or equal to 0.");
        }
        this.data.content.sizeInBytes = sizeInBytes;
        return this;
    };
    FileConsentCard.prototype.acceptContext = function (context) {
        this.data.content.acceptContext = context;
        return this;
    };
    FileConsentCard.prototype.declineContext = function (context) {
        this.data.content.declineContext = context;
        return this;
    };
    FileConsentCard.prototype.context = function (context) {
        this.acceptContext(context);
        this.declineContext(context);
        return this;
    };
    FileConsentCard.prototype.toAttachment = function () {
        return this.data;
    };
    return FileConsentCard;
}());
exports.FileConsentCard = FileConsentCard;
