// botbuilder-teams/lib/models/ListCard.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var o365ConnectorCard_1 = require("./o365ConnectorCard");
var ListCard = (function () {
    function ListCard(session) {
        this.session = session;
        this.data = {
            contentType: 'application/vnd.microsoft.teams.card.list',
            content: {}
        };
    }
    ListCard.prototype.title = function (text) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        if (text) {
            this.data.content.title = o365ConnectorCard_1.fmtText(this.session, text, args);
        }
        else {
            delete this.data.content.title;
        }
        return this;
    };
    ListCard.prototype.items = function (list) {
        var items = (list || []).map(ListCard.toListItem);
        this.data.content.items = items;
        return this;
    };
    ListCard.prototype.addItem = function (item) {
        var listItem = ListCard.toListItem(item);
        if (this.data.content.items) {
            var listItems = this.data.content.items;
            listItems.push(listItem);
        }
        else {
            this.data.content.items = [listItem];
        }
        return this;
    };
    ListCard.prototype.buttons = function (list) {
        var buttons = (list || []).map(function (button) {
            return (button.toAction) ?
                button.toAction() :
                button;
        });
        this.data.content.buttons = buttons;
        return this;
    };
    ListCard.prototype.toAttachment = function () {
        return this.data;
    };
    ListCard.toListItem = function (item) {
        return (item.toItem) ?
            item.toItem() :
            item;
    };
    return ListCard;
}());
exports.ListCard = ListCard;
var ListCardItem = (function () {
    function ListCardItem(session) {
        this.session = session;
        this.data = {
            type: ListCardItemType.resultItem,
        };
    }
    ListCardItem.prototype.type = function (type) {
        this.data.type = type;
        return this;
    };
    ListCardItem.prototype.title = function (text) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        if (text) {
            this.data.title = o365ConnectorCard_1.fmtText(this.session, text, args);
        }
        else {
            delete this.data.title;
        }
        return this;
    };
    ListCardItem.prototype.subtitle = function (text) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        if (text) {
            this.data.subtitle = o365ConnectorCard_1.fmtText(this.session, text, args);
        }
        else {
            delete this.data.subtitle;
        }
        return this;
    };
    ListCardItem.prototype.icon = function (url) {
        this.data.icon = url;
        return this;
    };
    ListCardItem.prototype.tap = function (action) {
        if (action) {
            this.data.tap = action.toAction ?
                action.toAction() :
                action;
        }
        else {
            delete this.data.tap;
        }
        return this;
    };
    ListCardItem.prototype.toItem = function () {
        return this.data;
    };
    return ListCardItem;
}());
exports.ListCardItem = ListCardItem;
var ListCardItemType;
(function (ListCardItemType) {
    ListCardItemType["resultItem"] = "resultItem";
    ListCardItemType["separator"] = "separator";
})(ListCardItemType = exports.ListCardItemType || (exports.ListCardItemType = {}));
