// botbuilder-teams/lib/models/TaskModuleAction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var builder = require("botbuilder");
var AdaptiveCard_1 = require("./AdaptiveCard");
var TaskModuleCardAction = (function () {
    function TaskModuleCardAction(session) {
        this.session = session;
        this.adaptorObj = new builder.CardAction(session);
        this.adaptorObj.type('invoke');
    }
    TaskModuleCardAction.prototype.title = function (text) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        var _a;
        (_a = this.adaptorObj).title.apply(_a, [text].concat(args));
        return this;
    };
    TaskModuleCardAction.prototype.value = function (v) {
        this.adaptorObj.value(v);
        return this;
    };
    TaskModuleCardAction.prototype.toAction = function () {
        var json = this.adaptorObj.toAction();
        var valJson = (typeof json.value === 'string') ? JSON.parse(json.value) : json.value;
        valJson.type = 'task/fetch';
        json.value = JSON.stringify(valJson);
        return json;
    };
    TaskModuleCardAction.prototype.toAdaptiveCardAction = function () {
        var btn = this.toAction();
        var adapterBtn = AdaptiveCard_1.AdaptiveCardBotBuilderAction.convertFromBotBuilderCardAction(btn);
        return adapterBtn;
    };
    return TaskModuleCardAction;
}());
exports.TaskModuleCardAction = TaskModuleCardAction;
