// botbuilder-teams/lib/TeamsChatConnector.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
var builder = require("botbuilder");
var msRest = require("ms-rest");
var RemoteQuery = require("./RemoteQuery/teams");
var RestClient = require("./RemoteQuery/RestClient");
var task = require("./models/TaskModuleResponse");
var WebResource = msRest.WebResource;
var TeamsChatConnector = (function (_super) {
    __extends(TeamsChatConnector, _super);
    function TeamsChatConnector(settings) {
        if (settings === void 0) { settings = {}; }
        var _this = _super.call(this, settings) || this;
        _this.queryHandlers = null;
        _this.allowedTenants = null;
        return _this;
    }
    TeamsChatConnector.prototype.fetchChannelList = function (serverUrl, teamId, callback) {
        var options = { customHeaders: {}, jar: false };
        var restClient = new RestClient(serverUrl, null);
        var remoteQuery = new RemoteQuery(restClient);
        this.getAccessToken(function (err, token) {
            if (!err && token) {
                options.customHeaders = {
                    'Authorization': 'Bearer ' + token
                };
                remoteQuery.fetchChannelList(teamId, options, callback);
            }
            else {
                return callback(new Error('Failed to authorize request'), null);
            }
        });
    };
    TeamsChatConnector.prototype.fetchTeamInfo = function (serverUrl, teamId, callback) {
        var options = { customHeaders: {}, jar: false };
        var restClient = new RestClient(serverUrl, null);
        var remoteQuery = new RemoteQuery(restClient);
        this.getAccessToken(function (err, token) {
            if (!err && token) {
                options.customHeaders = {
                    'Authorization': 'Bearer ' + token
                };
                remoteQuery.fetchTeamInfo(teamId, options, callback);
            }
            else {
                return callback(new Error('Failed to authorize request'), null);
            }
        });
    };
    TeamsChatConnector.prototype.fetchMembers = function (serverUrl, conversationId, callback) {
        var options = { customHeaders: {}, jar: false };
        var restClient = new RestClient(serverUrl, null);
        var remoteQuery = new RemoteQuery(restClient);
        this.getAccessToken(function (err, token) {
            if (!err && token) {
                options.customHeaders = {
                    'Authorization': 'Bearer ' + token
                };
                remoteQuery.fetchMemberList(conversationId, options, callback);
            }
            else {
                return callback(new Error('Failed to authorize request'), null);
            }
        });
    };
    TeamsChatConnector.prototype.fetchMembersWithPaging = function (serverUrl, conversationId) {
        var args = [];
        for (var _i = 2; _i < arguments.length; _i++) {
            args[_i - 2] = arguments[_i];
        }
        var pageSize = undefined;
        var continuationToken = undefined;
        var callback;
        var throwInvalidParameter = function () {
            throw new Error('Invalid parameters: ' + JSON.stringify(args));
        };
        if (!args) {
            throwInvalidParameter();
        }
        switch (args.length) {
            case 1:
                callback = args[0];
                break;
            case 2:
                if (typeof args[0] === 'string') {
                    continuationToken = args[0];
                }
                else if (typeof args[0] === 'number') {
                    pageSize = args[0];
                }
                else {
                    throwInvalidParameter();
                }
                callback = args[1];
                break;
            case 3:
                pageSize = args[0];
                continuationToken = args[1];
                callback = args[2];
                break;
            default:
                throwInvalidParameter();
        }
        var options = {
            pageSize: pageSize,
            continuationToken: continuationToken,
            customHeaders: {},
            jar: false
        };
        var restClient = new RestClient(serverUrl, null);
        var remoteQuery = new RemoteQuery(restClient);
        this.getAccessToken(function (err, token) {
            if (!err && token) {
                options.customHeaders = {
                    "Authorization": "Bearer " + token
                };
                remoteQuery.fetchMemberListWithPaging(conversationId, options, callback);
            }
            else {
                return callback(new Error("Failed to authorize request"), null);
            }
        });
    };
    TeamsChatConnector.prototype.startReplyChain = function (serverUrl, channelId, message, callback) {
        var options = { customHeaders: {}, jar: false };
        var restClient = new RestClient(serverUrl, null);
        var remoteQuery = new RemoteQuery(restClient);
        this.getAccessToken(function (err, token) {
            if (!err && token) {
                options.customHeaders = {
                    'Authorization': 'Bearer ' + token
                };
                var iMessage = null;
                if (message.toMessage) {
                    iMessage = message.toMessage();
                }
                else if (message.address) {
                    iMessage = message;
                }
                else {
                    throw new Error("Message type is wrong. Need either IMessage or IIsMessage");
                }
                var innerCallback = function (err, result) {
                    if (!callback) {
                        return;
                    }
                    if (result && result.hasOwnProperty("id") && result.hasOwnProperty("activityId")) {
                        var messageAddress = iMessage.address;
                        var address = __assign({}, messageAddress, { channelId: 'msteams', conversation: { id: result.id }, id: result.activityId });
                        if (address.user) {
                            delete address.user;
                        }
                        return callback(null, address);
                    }
                    else {
                        var error = new Error("Failed to start reply chain: no conversation ID and activity ID returned.");
                        return callback(error, null);
                    }
                };
                remoteQuery.beginReplyChainInChannel(channelId, iMessage, options, innerCallback);
            }
            else {
                if (callback) {
                    return callback(new Error('Failed to authorize request'), null);
                }
            }
        });
    };
    TeamsChatConnector.prototype.send = function (messages, done) {
        return _super.prototype.send.call(this, messages.filter(function (m) { return m.type !== "endOfConversation"; }), done);
    };
    TeamsChatConnector.prototype.setAllowedTenants = function (tenants) {
        if (tenants != null) {
            this.allowedTenants = tenants;
        }
    };
    TeamsChatConnector.prototype.resetAllowedTenants = function () {
        this.allowedTenants = null;
    };
    TeamsChatConnector.prototype.onO365ConnectorCardAction = function (handler) {
        this.o365CardActionHandler = handler;
    };
    TeamsChatConnector.prototype.onSigninStateVerification = function (handler) {
        this.signinStateVerificationHandler = handler;
    };
    TeamsChatConnector.prototype.onQuery = function (commandId, handler) {
        if (!this.queryHandlers) {
            this.queryHandlers = {};
        }
        this.queryHandlers[commandId] = handler;
    };
    TeamsChatConnector.prototype.onQuerySettingsUrl = function (handler) {
        this.querySettingsUrlHandler = handler;
    };
    TeamsChatConnector.prototype.onSettingsUpdate = function (handler) {
        this.settingsUpdateHandler = handler;
    };
    TeamsChatConnector.prototype.onSelectItem = function (handler) {
        this.selectItemInvokeHandler = handler;
    };
    TeamsChatConnector.prototype.onFileConsentCardResponse = function (handler) {
        this.fileConsentCardResponseHandler = handler;
    };
    TeamsChatConnector.prototype.onTaskModuleFetch = function (handler) {
        this.taskModuleFetchHandler = handler;
    };
    TeamsChatConnector.prototype.onTaskModuleSubmit = function (handler) {
        this.taskModuleSubmitHandler = handler;
    };
    TeamsChatConnector.prototype.onComposeExtensionFetchTask = function (handler) {
        this.composeExtensionFetchTaskHandler = handler;
    };
    TeamsChatConnector.prototype.onComposeExtensionSubmitAction = function (handler) {
        this.composeExtensionSubmitActionHandler = handler;
    };
    TeamsChatConnector.prototype.onAppBasedLinkQuery = function (handler) {
        this.appBasedLinkHandler = handler;
    };
    TeamsChatConnector.prototype.onDispatchEvents = function (events, callback) {
        if (this.allowedTenants) {
            var filteredEvents = [];
            for (var _i = 0, events_1 = events; _i < events_1.length; _i++) {
                var event = events_1[_i];
                if (event.sourceEvent.tenant && this.allowedTenants.indexOf(event.sourceEvent.tenant.id) > -1) {
                    filteredEvents.push(event);
                }
            }
            this.dispatchEventOrQuery(filteredEvents, callback);
        }
        else {
            this.dispatchEventOrQuery(events, callback);
        }
    };
    TeamsChatConnector.prototype.dispatchEventOrQuery = function (events, callback) {
        var realEvents = [];
        for (var _i = 0, events_2 = events; _i < events_2.length; _i++) {
            var event = events_2[_i];
            if (event.type === 'invoke') {
                var invoke = event;
                var invokeHandler = void 0;
                switch (invoke.name) {
                    case TeamsChatConnector.queryInvokeName:
                        if (this.queryHandlers) {
                            invokeHandler = this.dispatchQuery.bind(this);
                        }
                        break;
                    case TeamsChatConnector.querySettingUrlInvokeName:
                        if (this.querySettingsUrlHandler) {
                            invokeHandler = this.querySettingsUrlHandler.bind(this);
                        }
                        break;
                    case TeamsChatConnector.settingInvokeName:
                        if (this.settingsUpdateHandler) {
                            invokeHandler = this.settingsUpdateHandler.bind(this);
                        }
                        break;
                    case TeamsChatConnector.selectItemInvokeName:
                        if (this.selectItemInvokeHandler) {
                            invokeHandler = this.selectItemInvokeHandler.bind(this);
                        }
                        break;
                    case TeamsChatConnector.o365CardActionInvokeName:
                        if (this.o365CardActionHandler) {
                            invokeHandler = this.o365CardActionHandler.bind(this);
                        }
                        break;
                    case TeamsChatConnector.signinStateVerificationInvokeName:
                        if (this.signinStateVerificationHandler) {
                            invokeHandler = this.signinStateVerificationHandler.bind(this);
                        }
                        break;
                    case TeamsChatConnector.fileConsentInvokeName:
                        if (this.fileConsentCardResponseHandler) {
                            invokeHandler = this.fileConsentCardResponseHandler.bind(this);
                        }
                        break;
                    case TeamsChatConnector.taskModuleInvokeNameOfFetch:
                        if (this.taskModuleFetchHandler) {
                            invokeHandler = this.taskModuleFetchHandler.bind(this);
                        }
                        break;
                    case TeamsChatConnector.taskModuleInvokeNameOfSubmit:
                        if (this.taskModuleSubmitHandler) {
                            invokeHandler = this.taskModuleSubmitHandler.bind(this);
                        }
                        break;
                    case TeamsChatConnector.composeExtensionInvokeNameofFetchTask:
                        if (this.composeExtensionFetchTaskHandler) {
                            invokeHandler = this.composeExtensionFetchTaskHandler.bind(this);
                        }
                        break;
                    case TeamsChatConnector.composeExtensionInvokeNameofSubmitAction:
                        if (this.composeExtensionSubmitActionHandler) {
                            invokeHandler = this.composeExtensionSubmitActionHandler.bind(this);
                        }
                        break;
                    case TeamsChatConnector.appBasedLinkInvokeName:
                        if (this.appBasedLinkHandler) {
                            invokeHandler = this.appBasedLinkHandler.bind(this);
                        }
                        break;
                    default:
                        realEvents.push(event);
                        break;
                }
                if (invokeHandler) {
                    try {
                        invokeHandler(invoke, invoke.value, callback);
                    }
                    catch (e) {
                        return callback(e, null, 500);
                    }
                }
                else {
                    realEvents.push(event);
                }
            }
            else {
                realEvents.push(event);
            }
        }
        if (realEvents.length > 0) {
            _super.prototype.onDispatchEvents.call(this, realEvents, callback);
        }
    };
    TeamsChatConnector.prototype.dispatchQuery = function (event, query, callback) {
        var handler = this.queryHandlers[query.commandId];
        if (handler) {
            handler(event, query, callback);
        }
        else {
            return callback(new Error("Query handler [" + query.commandId + "] not found."), null, 500);
        }
    };
    TeamsChatConnector.o365CardActionInvokeName = 'actionableMessage/executeAction';
    TeamsChatConnector.signinStateVerificationInvokeName = 'signin/verifyState';
    TeamsChatConnector.queryInvokeName = 'composeExtension/query';
    TeamsChatConnector.querySettingUrlInvokeName = 'composeExtension/querySettingUrl';
    TeamsChatConnector.selectItemInvokeName = 'composeExtension/selectItem';
    TeamsChatConnector.settingInvokeName = 'composeExtension/setting';
    TeamsChatConnector.fileConsentInvokeName = 'fileConsent/invoke';
    TeamsChatConnector.taskModuleInvokeNameOfFetch = task.taskModuleInvokeNameOfFetch;
    TeamsChatConnector.taskModuleInvokeNameOfSubmit = task.taskModuleInvokeNameOfSubmit;
    TeamsChatConnector.composeExtensionInvokeNameofFetchTask = 'composeExtension/fetchTask';
    TeamsChatConnector.composeExtensionInvokeNameofSubmitAction = 'composeExtension/submitAction';
    TeamsChatConnector.appBasedLinkInvokeName = 'composeExtension/queryLink';
    return TeamsChatConnector;
}(builder.ChatConnector));
exports.TeamsChatConnector = TeamsChatConnector;
